hotspots: Copy hotspots from other accounts with the same email.

mrchntia opened the pull request:
Fixes this:
https://chat.zulip.org/#narrow/stream/9-issues/topic/disable.20helpful.20popup

- If a new user signs up, the hotspots are copied from user profiles that are registered with the same email, so they are not shown to the user a second time.
- If the user selects a user profile to copy the settings from only the viewed hotspots of this user are copied even if there are other profiles with the same email that have seen more hotspots.

The tutorial status is only updated if at least one hotspot is updated, this assumes that the tutorial status is never changed independently of hotspots.
Is this the desired behavior?

**Testing plan:** <!-- How have you tested? -->
The function test_signup_without_user_settings_from_another_realm and test_signup_with_user_settings_from_another_realm test the  behavior. Originaly I created another new test method for this feature, but it ended up being the same code like in test_signup_without_user_settings_from_another_realm so I removed it again.
Manually tested by running a server and creating different users in different realms.

hackerkid reviewed (commented):
@mrchntia Thanks for working on this. There are two ways to implement this.

1. Don't show hotspots if the user already has an existing account irrespective of whether the user choose to copy the settings during sign up.

2. Copy only the hotspots that the user has not seen yet irrespective of whether the user has decided to copy the settings or not during the sign up.

You have implemented #2 with an additional change of only copying the hotspots from the source profile if the user selects an existing account.


The issue with #2 is that, lets say the user created an account and saw 2/5 hotspots. Then the user went on to create another account. So we would show the remaining 3 hotspots to the user in his second account during onboadring. If the user goes back to the first account again, we should again show the 3 hotspots. I think one way we can avoid this would be to update `get_next_hotspots` to take in account of all profiles while fetching `seen_hotspots`. In that way we don't have to copy the hotspot state from exiting accounts during sign up. Plus `UserHotspot` table would be accurate. It will keep track of exactly which all hotspots were seen by which all profiles. Can you update the code to implement it this way? Also let me know if there any further queries.


Also mentioning https://zulip.readthedocs.io/en/latest/subsystems/hotspots.html in case you are not aware of it. It has some useful information on hotspots in general.

mrchntia commented on zerver/lib/create_user.py:
```diff
@@ -43,7 +43,8 @@ def copy_user_settings(source_profile: UserProfile, target_profile: UserProfile)
 
-    copy_hotpots(source_profile, target_profile)
+    copy_hotspots(source_profile, target_profile)
+    target_profile.onboarding_steps = source_profile.onboarding_steps
```
You are right it should be inside copy_hotspots and save is missing.

mrchntia commented:
@hackerkid Thanks a lot for this review! Your suggestion makes sense, I will update the code to implement this.

Thanks for the hint to the documentation, I already read that. I'm still wondering about the onboarding_steps, I can't find there usage in the code other than in the function copy_hotsots. So, when will they be updated and are the onboarding_steps representing the hotspots?

hackerkid commented:
@mrchntia I don't think we make use of `onboarding_steps` anywhere. So we can just ignore it.

mrchntia commented:
I implemented the suggestion and added tests for it. I also deleted the function copy_hotspots because it is not needed anymore.
I deleted the first commit, not sure if this was necessary.
@hackerkid could you review the changes?

hackerkid commented on zerver/tests/test_signup.py:
```diff
@@ -3074,7 +3073,6 @@ def test_signup_without_user_settings_from_another_realm(self) -> None:
         self.assertEqual(hamlet.high_contrast_mode, False)
         self.assertEqual(hamlet.enable_stream_audible_notifications, False)
         self.assertEqual(hamlet.enter_sends, False)
-        self.assertEqual(hamlet.tutorial_status, UserProfile.TUTORIAL_WAITING)
```
Actually don't delete this. Just update these tests to check the appropriate value if necessary.

hackerkid commented on zerver/tests/test_users.py:
```diff
@@ -1174,9 +1167,6 @@ def test_copy_user_settings(self) -> None:
         self.assertEqual(cordelia.enter_sends, False)
         self.assertEqual(hamlet.enter_sends, True)
 
-        hotspots = list(UserHotspot.objects.filter(user=iago).values_list('hotspot', flat=True))
-        self.assertEqual(hotspots, hotspots_completed)
```
Same here as well. Don't delete them. Just check for the expected value.

hackerkid commented on zerver/lib/hotspots.py:
```diff
@@ -75,12 +75,3 @@ def get_next_hotspots(user: UserProfile) -> List[Dict[str, object]]:
     user.save(update_fields=['tutorial_status'])
     return []
-
-def copy_hotpots(source_profile: UserProfile, target_profile: UserProfile) -> None:
-    for userhotspot in frozenset(UserHotspot.objects.filter(user=source_profile)):
-        UserHotspot.objects.create(user=target_profile, hotspot=userhotspot.hotspot,
-                                   timestamp=userhotspot.timestamp)
-
-    target_profile.tutorial_status = source_profile.tutorial_status
-    target_profile.onboarding_steps = source_profile.onboarding_steps
-    target_profile.save(update_fields=['tutorial_status', 'onboarding_steps'])
```
All the changes in this commit should go along with the first commit.

The reason why we don't want a separate commit for removing `copy_hotpots` is that if we merge only the first commit, the system would be behaving incorrectly. That is `get_next_hotspots` would be fetching the hotspots from other profiles and `copy_hotpots` would be used used for copying the hotspots. We don't want that.  So both the changes should go in a single commit.

In general, whenever you create a commit x, ensure that the system works correctly if we deploy only the commits till x to production (even if the PR has many other commits after x).

mrchntia replied:
Thanks for the explanation, this makes sense.

hackerkid reviewed (commented):
@mrchntia The changes look great. Some minor comments.

mrchntia commented:
Sorry for the late reply, I didn't find time to work over Christmas.
Thanks for the review. I'll do the changes and update the PR. :)

mrchntia commented on zerver/tests/test_hotspots.py:
```diff
@@ -41,6 +41,87 @@ def test_send_all(self) -> None:
         with self.settings(DEVELOPMENT=True, ALWAYS_SEND_ALL_HOTSPOTS = True):
             self.assertEqual(len(ALL_HOTSPOTS), len(get_next_hotspots(self.user)))
 
+    def test_two_users_with_same_email_all_done(self) -> None:
+        for hotspot in ALL_HOTSPOTS:
+            do_mark_hotspot_as_read(self.user, hotspot)
+
+        user_in_red = do_create_user(
+            'user@zulip.com',
+            'password',
+            do_create_realm("red", "red"),
+            'user',
+        )
+
+        self.assertEqual(get_next_hotspots(user_in_red), [])
+
+    def test_two_users_with_same_email_some_done_some_not(self) -> None:
+        for hotspot in ["intro_reply", "intro_streams"]:
+            do_mark_hotspot_as_read(self.user, hotspot)
+
+        user_in_red = do_create_user(
+            'user@zulip.com',
+            'password',
+            do_create_realm("red", "red"),
+            'user',
+        )
+
+        hotspots = get_next_hotspots(user_in_red)
+        self.assertEqual(len(hotspots), 1)
+        self.assertEqual(hotspots[0]['name'], 'intro_topics')
```
Ok, I'll `delete test_three_users_with_same_email_all_done` and `test_three_users_with_same_email_some_done_some_not`.

hackerkid replied:
I think we can delete the tests for two users and keep the ones with 3 users since that covers this as well.

mrchntia commented on zerver/tests/test_hotspots.py:
```diff
@@ -41,6 +41,87 @@ def test_send_all(self) -> None:
+            do_create_realm("red", "red"),
+            'user',
+        )
+
```
Do I even need this since I'll delete this function anyway?

mrchntia replied:
Ok, since I keep this function, I am not exactly sure what you mean with "add a check for `get_next_hotspots`?

hackerkid replied:
I just meant to add a check that make sure that the `get_next_hotspots` return the expected value for `self.user` right after we mark two hotspots as read.

mrchntia replied:
I updated the code, I hope this is what you mean. If not I am happy to change this again.

hackerkid commented on zerver/tests/test_hotspots.py:
```diff
@@ -41,6 +41,57 @@ def test_send_all(self) -> None:
+        hotspots = get_next_hotspots(user_in_blue)
+        self.assertEqual(len(hotspots), 1)
+        self.assertEqual(hotspots[0]['name'], 'intro_gear')
+
```
I think we can also mark intro_gear as read after this and then check for get_next_hotspots for the all 3 users again?

mrchntia commented:
It seems like the test is failing for an unknown reason that is not related to my code. This PR https://github.com/zulip/zulip/pull/16997 will probably fix it.

mrchntia commented:
@hackerkid The failing frontend test is now fix and does not fail anymore. I ran `./tools/test-all` locally and all tests are passed.
Because I didn't change anything on my branch force push only tells me that everyting is up-to-date therefore it doesn't run the CI tests again. Is there an other way to run the CI tests again so it shows that all checks are successful?

sahil839 commented:
You can do this https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/error.20checking.20links.20CI/near/825573 to repush and trigger CI again.
